### PR TITLE
chore(api): support ignore password if login source type is LDAP for creating user API

### PR DESCRIPTION
**Pull request by appleboy**
- Modify the `Password` field in `CreateUserOption` struct to remove the `Required` tag
- Update the `v1_json.tmpl` template to include the `email` field and remove the `password` field

### CODE REVIEW

**Comment by lunny** [comment]
Maybe we need to check whether the user is a plain type or other type?

**Comment by appleboy** [comment]
We only support login type as LDAP so we need to ignore the password field for checking the request body.

**Comment by appleboy** [comment]
You can see the admin page for creating users in LDAP source type.

image

**Comment by appleboy** [comment]
@lunny Add checking in Plain type.

File: routers/api/v1/admin/user.go [hunk 1]
```diff
@@ -93,6 +93,13 @@ func CreateUser(ctx *context.APIContext) {
...
 		return
 	}
+
+	if u.LoginType == auth.Plain && len(form.Password) == 0 {
+		err := errors.New("PasswordIsRequired")
+		ctx.Error(http.StatusBadRequest, "PasswordIsRequired", err)
+		return
+	}
+
 	if !password.IsComplexEnough(form.Password) {
```
Comment (lunny) [inline routers/api/v1/admin/user.go]: Should this be changed according login type?

Comment (appleboy) [reply routers/api/v1/admin/user.go]: Yes. We can refer to the following source code on the admin page.

https://github.com/go-gitea/gitea/blob/9a15267871dcb1f1f8b39115b8e68ed1f42bf21d/routers/web/admin/users.go#L154-L179

Comment (appleboy) [reply routers/api/v1/admin/user.go]: @wolfogre See the commit [f9933a1](https://github.com/go-gitea/gitea/pull/28491/commits/f9933a18159c24d76036fab7e872794d43479437)

Comment (lunny) [reply routers/api/v1/admin/user.go]: I think it's not enough.

Comment (appleboy) [reply routers/api/v1/admin/user.go]: @lunny See [dea9c50](https://github.com/go-gitea/gitea/pull/28491/commits/dea9c50baaac294a6c10ffd2317583980893c290)

**Review by lunny — APPROVED** [review]

**Comment by appleboy** [comment]
@lunny Can we backport the PR to the 1.21.3 version?

**Comment by wxiaoguang** [comment]
Could you elaborate the reason why it needs to "create a user without password"?

If you are using LDAP, I think you could just "sync" the LDAP users into Gitea, then no need to call the "create user API"?

**Comment by appleboy** [comment]
@wolfogre 

Our company is currently in the process of migrating from Bitbucket to Gitea service. Since Bitbucket is integrated with LDAP authentication, there are 20,000 users in LDAP. However, we only want to migrate the users who are currently using Bitbucket to Gitea, rather than transferring all users.

I have developed a migration tool that, during the project conversion, adds users based on project permissions. Therefore, it requires the use of the "add user" API. The login source will be filled with LDAP, and when using LDAP, there is no need to provide a password.

https://github.com/appleboy/BitbucketServer2Gitea

**Review by wolfogre — APPROVED** [review]

**Comment by lunny** [comment]
> @wxiaoguang
> 
> Our company is currently in the process of migrating from Bitbucket to Gitea service. Since Bitbucket is integrated with LDAP authentication, there are 20,000 users in LDAP. However, we only want to migrate the users who are currently using Bitbucket to Gitea, rather than transferring all users.
> 
> I have developed a migration tool that, during the project conversion, adds users based on project permissions. Therefore, it requires the use of the "add user" API. The login source will be filled with LDAP, and when using LDAP, there is no need to provide a password.
> 
> https://github.com/appleboy/BitbucketServer2Gitea

BTW: any chance to merge the logic into Gitea? So we can have a new migrate from bitbucket like others.

**Pull request merged**